Adding to_string methods

not-Ian opened the pull request:
Adding to_string methods
Also added some build items to .gitignore 

Re: #3451

not-Ian committed:
adding method

not-Ian committed:
telemetry_req, telemetry_ack

not-Ian committed:
adding to_string to frontier_req

dsiganos commented on nano/node/common.cpp:
```diff
@@ -696,6 +696,21 @@ void nano::keepalive::serialize (nano::stream & stream_a) const
+
+	stream << "\n" + header.to_string () + "\n";
+
+	for (auto peer (peers.begin ()), j (peers.end ()); peer != j; ++peer)
```
Use:
```
 auto peer = peers.begin ()
 ```

not-Ian replied:
done. Should I make this change in other places in the same file where they use `auto peer (peers.begin ())` as opposed to `auto peer = peers.begin ()`? Like, in `nano::keepalive::serialize` around line 690

dsiganos replied:
In general yes, but for now since this is your first PR, let's just focus on getting it through the review process and merged.

dsiganos commented on nano/node/common.cpp:
```diff
@@ -696,6 +696,21 @@ void nano::keepalive::serialize (nano::stream & stream_a) const
+{
+	std::stringstream stream;
+
+	stream << "\n" + header.to_string () + "\n";
```
Why is there a leading new line? That looks weird. It should be up to the caller to put leading or trailing whitespace. This code should use minimal whitespace.

not-Ian replied:
done

dsiganos commented on nano/node/common.cpp:
```diff
@@ -696,6 +696,21 @@ void nano::keepalive::serialize (nano::stream & stream_a) const
 	}
 }
 
+std::string nano::keepalive::to_string ()
```
This function can be const.

not-Ian replied:
marked const in most recent commit

dsiganos commented on .gitignore:
```diff
@@ -57,6 +57,9 @@ Release
 *.aps
 *.db
 *.opendb
+/out/**
```
Let's not commit changes to .gitingore that are not relevant to this PR.

not-Ian replied:
I have deleted these changes

dsiganos commented on nano/node/common.cpp:
```diff
@@ -996,6 +1011,17 @@ bool nano::frontier_req::operator== (nano::frontier_req const & other_a) const
 	return start == other_a.start && age == other_a.age && count == other_a.count;
 }
 
+std::string nano::frontier_req::to_string ()
```
remove leading space, this could be compressed to one line without whitespace if you shorted the message.

E.g.
```
start=xxxx age=yyyy count=zzzz
```

Try not to avoid whitespace, ideally use bio vertical whitespace.
Use vertical whitespace only where it adds a lot of value.

not-Ian replied:
pushed just now

dsiganos commented on nano/node/common.cpp:
```diff
@@ -1189,6 +1215,15 @@ void nano::telemetry_req::visit (nano::message_visitor & visitor_a) const
 	visitor_a.telemetry_req (*this);
 }
 
+std::string nano::telemetry_req::to_string () 
```
This could be a one-liner
```
return header.to_string()
```

dsiganos commented on nano/node/common.cpp:
```diff
@@ -1221,6 +1256,15 @@ void nano::telemetry_ack::serialize (nano::stream & stream_a) const
+std::string nano::telemetry_ack::to_string () const
+{
+	std::stringstream stream;
+
```
could be a one-liner

not-Ian replied:
are you thinking just drop the stringstream object entirely? I'm only using it because I saw it was used in a separate to_string function. Is it only useful if you're formatting multiple variables together or something?

not-Ian replied:
I also made telemetry_req::to_string() a one liner. It was very similar

dsiganos replied:
string stream is a flexible stream buffer that can accumulate lots of data in a flexible way and you can output formatting rules to format your data. It is convenient.
But in this case, your data is ready, no processing needed, just pass it along as it is.

dsiganos commented on nano/node/common.hpp:
```diff
@@ -424,6 +426,7 @@ class telemetry_req final : public message
 	void serialize (nano::stream &) const override;
 	bool deserialize (nano::stream &);
 	void visit (nano::message_visitor &) const override;
+	std::string to_string ();
```
this can be marked as const

not-Ian replied:
marked const in most recent commit

dsiganos commented on nano/node/common.hpp:
```diff
@@ -368,6 +369,7 @@ class frontier_req final : public message
 	uint32_t age;
 	uint32_t count;
 	static std::size_t constexpr size = sizeof (start) + sizeof (age) + sizeof (count);
+	std::string to_string ();
```
this can be marked as const

not-Ian replied:
marked const in most recent commit

dsiganos commented on nano/node/common.hpp:
```diff
@@ -311,6 +311,7 @@ class keepalive final : public message
 	bool operator== (nano::keepalive const &) const;
 	std::array<nano::endpoint, 8> peers;
 	static std::size_t constexpr size = 8 * (16 + 2);
+	std::string to_string ();
```
this can be marked as const

not-Ian replied:
marked const in most recent commit

dsiganos commented on nano/node/common.cpp:
```diff
@@ -1189,6 +1215,15 @@ void nano::telemetry_req::visit (nano::message_visitor & visitor_a) const
 	visitor_a.telemetry_req (*this);
 }
 
+std::string nano::telemetry_req::to_string () 
```
this can be marked as const

not-Ian replied:
marked const in most recent commit

not-Ian committed:
making some functions const that can be const
removing my changes in .gitignore

not-Ian committed:
added confirm_req::to_string

dsiganos commented on nano/node/common.cpp:
```diff
@@ -696,6 +696,21 @@ void nano::keepalive::serialize (nano::stream & stream_a) const
+
+	stream << header.to_string () + "\n";
+
+	for (auto peer = peers.begin (), j (peers.end ()); peer != j; ++peer)
```
Please use
```
j = peers.end()
```

we prefer the assignment style of initialisation

not-Ian replied:
if I see somewhere else that's not doing the assignment style of initialization, can I change it, or would that be out of the scope of this PR and not approved of

dsiganos replied:
You can make small obvious changes but bigger changes should be done in a separate PR

not-Ian commented:
Example outputs:

nano::confirm_req:

`NetID: 5243(live), VerMaxUsingMin: 18/18/18, MsgType: 4(confirm_req), Extensions: 1100`

---

nano::telemetry_ack:

```
{
    "block_count": "0",
    "cemented_count": "0",
    "unchecked_count": "0",
    "account_count": "0",
    "bandwidth_cap": "0",
    "peer_count": "0",
    "protocol_version": "0",
    "uptime": "0",
    "genesis_block": "0000000000000000000000000000000000000000000000000000000000000000",
    "major_version": "0",
    "minor_version": "0",
    "patch_version": "0",
    "pre_release_version": "0",
    "maker": "0",
    "timestamp": "0",
    "active_difficulty": "0000000000000000"
}
```

---

nano::frontier_req:

`Start account: 0000000000000000000000000000000000000000000000000000000000000000 | Maximum age of account: 1647681972 | Maximum number of accounts to include: 1048576`

---

nano::confirm_ack:

`Account: A30E0A32ED41C8607AA9212843392E853FCBCB4E7CB194E35C94F07F91DE59EF`

---

nano::telemetry_req:

`NetID: 5243(live), VerMaxUsingMin: 19/19/18, MsgType: 12(telemetry_req), Extensions: 0000`

not-Ian committed:
added confirm_ack to_string, fixed frontier_req to_string

not-Ian committed:
added to_string in node_id_handshake

not-Ian committed:
created the final 5 to_string functions

not-Ian commented:
And the rest of them:

node::node_id_handshake:

```
NetID: 5243(live), VerMaxUsingMin: 18/18/18, MsgType: 10(node_id_handshake), Extensions: 0003
Query cookie: 794DE15DC2B3CB84CE0AF3917F5B3A29C5219F83BD0B8BDB84A293C064716C18
Account: 1A2C0156415D1386FB86D03CD99966E73C1DBBC0985D9D86C484F7C554D00308 | 349660F5768BA07E479C76CE5F4E44189325A84F9A2DE6CF23E4AF9D40D7A98344589EAAC3A738D1240901B77C2AE35A063A8AA1E056DA7E22D1BD365DD0CC0D
```

---

node::publish

```
{
    "type": "state",
    "account": "nano_1chan9rmpx8wxh83wcmsxf4ywcaqq5riimq8d3876iouzdjk9nsn4cf56qn4",
    "previous": "67950A8DD4B95F3E201885DF54FCF3CDDFBA572F8332E3EDE6D4CB39A3E89DF1",
    "representative": "nano_3bsnis6ha3m9cepuaywskn9jykdggxcu8mxsp76yc3oinrt3n7gi77xiggtm",
    "balance": "99858859438706100489898610573647",
    "link": "8D4B2671F35B03656D97881E75878D8E0BF374ACC61A92FD04735F6DFE4E097A",
    "link_as_account": "nano_35cd6srz8pr5eopsh41ygp5ru5idyftcsjitkdyiawtzfqz6w4dth7338pr7",
    "signature": "0A605F86F14D0EA9A475F614DAEC4E270FE9FB38229F7332F86184B00008DCAADEB97BFFE0001D68AA3C4F3466B0F6305CD67D6DBB01F5D30DD9F25ABB6BF403",
    "work": "24578869cb8c5048"
}
```

---

node::bulk_pull:

`Start account: 03489141285CDF3936090F1F6E68238FEA79F419E5C4F15C3E969A7E26B28DC2 | End block hash: 74CD5BBA03C45481424183F2E68FA812FEAD481D5D5B509B4B576FEBDBA52B1C | Count: 0`

---

node::bulk_pull_account:

`Account: FD98692F080CED721C8449E034F3D8ADC11C890CE46B208D8F3CC73A765D54EB | Minimum amount: 000000000000D3C21BCECCEDA1000000 | Pending hash and amount`

---

node::bulk_push:

`NetID: 5243(live), VerMaxUsingMin: 19/19/18, MsgType: 7(bulk_push), Extensions: 0000NetID: 5243(live), VerMaxUsingMin: 19/19/18, MsgType: 7(bulk_push), Extensions: 0000Socket error: connect, 10061`

dsiganos commented:
It is good to keep logs of this nature as concise as possible.
I would change this:
```
Start account: 0000000000000000000000000000000000000000000000000000000000000000 | Maximum age of account: 1647681972 | Maximum number of accounts to include: 1048576
```

to this:
```
0000000000000000000000000000000000000000000000000000000000000000 maxage=1647681972 count=1048576
```

not-Ian committed:
adding break statements

not-Ian committed:
minimizing print statements

not-Ian committed:
minamt to min, added a space

not-Ian committed:
made this a one liner

not-Ian commented:
gonna work on tests next. And also clean up the two functions that are returning JSON

not-Ian committed:
changing this method slightly to be shorter

not-Ian committed:
basic first test structure. Running into issues with visual studio identifying a target I'm trying to make that will only run one test

not-Ian committed:
needed to fix the `vote_cache.cpp` test. Also made a very basic version of the keepalive::to_string () function test

dsiganos committed:
Do not put a new line at the end of log

That should be the choice of the caller, if it should be terminated in new
line or not.

dsiganos committed:
Example of how to add peers into keepalive class

dsiganos committed:
Run ci/clang-format-do.sh to fix formatting

not-Ian committed:
added some tests, removed new line at end of log for confirm_req

not-Ian committed:
removed new lines at end of strings

not-Ian committed:
Merge branch 'adding-to-string-methods' of github.com:not-Ian/nano-node into adding-to-string-methods

# Conflicts:
#	nano/core_test/common.cpp
#	nano/node/common.cpp

not-Ian committed:
final tests, improving logs slightly re: trailing new lines

not-Ian committed:
Merge branch 'develop' of github.com:nanocurrency/nano-node into adding-to-string-methods

# Conflicts:
#	nano/node/common.cpp
#	nano/node/common.hpp

not-Ian committed:
moving tests

not-Ian committed:
pushing the keepalive hpp change, must have missed it earlier

dsiganos committed:
Fix compilation problems

dsiganos commented on nano/node/messages.cpp:
```diff
@@ -707,6 +707,21 @@ bool nano::keepalive::deserialize (nano::stream & stream_a)
+
+	stream << header.to_string ();
+
+	for (auto peer = peers.begin (), peers_end = peers.end (); peer != peers_end; ++peer)
```
There is no need to take a copy of `peers_end = peers.end ();` to avoid calling the function repeatedly. compilers know how to optimise these things easily. The code should be written for readability first, not for optimisation.

This is better because it is easily readable and because the compiler will make the optimisation anyway:
```
for (auto peer = peers.begin (); peer != peers.end(); ++peer)
```

dsiganos commented on nano/slow_test/vote_cache.cpp:
```diff
@@ -215,7 +215,7 @@ TEST (vote_cache, perf_multithreaded)
 	node.active.clear ();
 
-	run_parallel (thread_count, [&node, &reps, &blocks] (int index) {
+	run_parallel (thread_count, [&node, &reps, &blocks, &vote_count, &single_vote_size, &single_vote_reps] (int index) {
```
This seems unrelated.

not-Ian replied:
I originally needed this change to get tests to run

dsiganos replied:
OK

dsiganos commented on nano/node/messages.cpp:
```diff
@@ -1581,4 +1680,20 @@ std::size_t nano::node_id_handshake::size (nano::message_header const & header_a
+	std::stringstream stream;
+
+	stream << header.to_string () + "\n";
+	stream << "cookie=" + query->to_string ();
```
The query is optional, so you need to check it has a value before trying to access it.

dsiganos commented on nano/node/messages.cpp:
```diff
@@ -1287,6 +1381,11 @@ bool nano::telemetry_ack::is_empty_payload () const
 
+std::string nano::telemetry_ack::to_string () const
+{
+	return data.to_string ();
```
It is possible for telemetry data to have empty payload. You need to check is_empty_payload.

dsiganos commented on nano/core_test/message.cpp:
```diff
@@ -276,3 +276,242 @@ TEST (message, bulk_pull_serialization)
+
+TEST (message, keepalive_to_string)
+{
+	nano::work_thresholds work_threshold = nano::work_thresholds (0, 0, 0);
```
You can create the keepalive object in one line like this:
```
nano::keepalive keepalive = nano::keepalive (nano::dev::network_params.network);
```

not-Ian commented:
Closing in favor of smaller, individual PRs
